File: common/src/main/java/generations/gg/generations/core/generationscore/world/level/block/entities/shrines/altar/TimeSpaceAltarBlockEntity.java
package generations.gg.generations.core.generationscore.world.level.block.entities.shrines.altar;

import earth.terrarium.botarium.common.item.ItemContainerBlock;
import earth.terrarium.botarium.common.item.SerializableContainer;
import generations.gg.generations.core.generationscore.GenerationsCore;
import generations.gg.generations.core.generationscore.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.util.ExtendedsimpleItemContainer;
import generations.gg.generations.core.generationscore.util.GenerationsUtils;
import generations.gg.generations.core.generationscore.world.entity.block.PokemonUtil;
import generations.gg.generations.core.generationscore.world.item.legends.CreationTrioItem;
import generations.gg.generations.core.generationscore.world.item.GenerationsItems;
import generations.gg.generations.core.generationscore.world.item.legends.RedChainItem;
import generations.gg.generations.core.generationscore.world.level.block.entities.GenerationsBlockEntities;
import generations.gg.generations.core.generationscore.world.level.block.entities.shrines.InteractShrineBlockEntity;
import generations.gg.generations.core.generationscore.world.level.block.entities.shrines.ShrineBlockEntity;
import generations.gg.generations.core.generationscore.world.level.block.shrines.ShrineBlock;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.block.state.BlockState;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class TimeSpaceAltarBlockEntity extends ShrineBlockEntity implements ItemContainerBlock, ModelContextProviders.VariantProvider {

    private TimeSpaceAltarItemStackHandler handler;

    public TimeSpaceAltarBlockEntity(BlockPos pos, BlockState state) {
        super(GenerationsBlockEntities.TIMESPACE_ALTAR.get(), pos, state);
    }


    @Override
    public String getVariant() {
        return handler.hasRedChain() ? "red chain" : "none";
    }

    @Override
    public TimeSpaceAltarItemStackHandler getContainer() {
        return handler == null ? (handler = new TimeSpaceAltarItemStackHandler()) : handler;
    }

    @Nullable
    public CreationTrioItem getOrb() {
        if (handler.getItem(0).getItem() instanceof CreationTrioItem item) return item;
        return null;
    }

    public class TimeSpaceAltarItemStackHandler extends ExtendedsimpleItemContainer {

        public TimeSpaceAltarItemStackHandler() {
            super(TimeSpaceAltarBlockEntity.this, 2);
        }

        @Override
        public boolean isItemValid(int slot, @NotNull ItemStack stack) {
            if (slot == 0) {
                return stack.getItem() instanceof CreationTrioItem;
            } else {
                return stack.is(GenerationsItems.RED_CHAIN.get()) && RedChainItem.isEnchanted(stack);
            }
        }

        @Override
        protected int getStackLimit(int slot, @NotNull ItemStack stack) {
            return 1;
        }

        public boolean hasRedChain() {
            return !getItem(1).isEmpty();
        }

        public boolean hasOrb(ServerPlayer player) {
            var orb = getItem(0);

            return !orb.isEmpty() && !GenerationsCore.CONFIG.caught.capped(player, ((CreationTrioItem) orb.getItem()).getSpeciesId());
        }


        public ItemStack extractItem() {
            for (int i = 0; i < 2; i++)
                if (!getItem(i).isEmpty()) return extractItem(i, 1, false);

            return ItemStack.EMPTY;
        }

        public boolean shouldSpawn(ServerPlayer player) {
            return hasRedChain() && hasOrb(player);
        }

        public void dumpAllIntoPlayerInventory(ServerPlayer player) {
            for (ItemStack stack : getItems())
                player.getInventory().placeItemBackInInventory(stack);

            clearContent();
        }
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/entities/shrines/altar/TimeSpaceAltarBlockEntity.kt
package generations.gg.generations.core.generationscore.common.world.level.block.entities.shrines.altar

import earth.terrarium.common_storage_lib.item.impl.vanilla.VanillaDelegatingSlot
import earth.terrarium.common_storage_lib.item.impl.vanilla.WrappedVanillaContainer
import earth.terrarium.common_storage_lib.item.util.ItemProvider
import earth.terrarium.common_storage_lib.resources.item.ItemResource
import earth.terrarium.common_storage_lib.storage.base.CommonStorage
import generations.gg.generations.core.generationscore.common.GenerationsCore
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.instanceOrNull
import generations.gg.generations.core.generationscore.common.util.ExtendedsimpleItemContainer
import generations.gg.generations.core.generationscore.common.util.asResource
import generations.gg.generations.core.generationscore.common.world.item.GenerationsItems
import generations.gg.generations.core.generationscore.common.world.item.components.GenerationsDataComponents
import generations.gg.generations.core.generationscore.common.world.item.legends.CreationTrioItem
import generations.gg.generations.core.generationscore.common.world.level.block.entities.GenerationsBlockEntities
import generations.gg.generations.core.generationscore.common.world.level.block.entities.shrines.InteractShrineBlockEntity
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.block.state.BlockState

class TimeSpaceAltarBlockEntity(pos: BlockPos, state: BlockState) : InteractShrineBlockEntity(GenerationsBlockEntities.TIMESPACE_ALTAR, pos, state), ItemProvider.BlockEntity, ModelContextProviders.VariantProvider {
    private var handler: TimeSpaceAltarItemStackHandler = TimeSpaceAltarItemStackHandler()

    override fun getVariant(): String {
        return if (container.hasRedChain()) "red chain" else "none"
    }

    val container: TimeSpaceAltarItemStackHandler
        get() = handler

    val orb: CreationTrioItem?
        get() = container.getItem(0).item.instanceOrNull<CreationTrioItem>()



    inner class TimeSpaceAltarItemStackHandler : ExtendedsimpleItemContainer(2) {
        override fun canPlaceItem(slot: Int, stack: ItemStack): Boolean {
            return if (slot == 0) {
                stack.item is CreationTrioItem
            } else {
                stack.`is`(GenerationsItems.RED_CHAIN.value()) && stack.get(GenerationsDataComponents.ENCHANTED.value()) == false
            }
        }

        override fun getMaxStackSize(): Int {
            return 1
        }


        fun hasRedChain(): Boolean {
            return !this.getItem(1).isEmpty
        }

        fun hasOrb(player: ServerPlayer?): Boolean {
            val orb = this.getItem(0)

            return !orb.isEmpty && GenerationsCore.CONFIG!!.caught.capped(
                player,
                (orb.item as CreationTrioItem).speciesId
            )
        }

        fun extract(index: Int, simulate: Boolean): ItemStack { //TODO: This probably really jank and I need to change this completley.
            val src = getItem(index);
            val stack = src
            setItem(index, ItemStack.EMPTY)

            return stack
        }

        fun extractItem(): ItemStack {
            for (i in 0..1) if (!this.getItem(i).isEmpty) return extract(1, false)

            return ItemStack.EMPTY
        }

        fun shouldSpawn(player: ServerPlayer): Boolean {
            return hasRedChain() && hasOrb(player)
        }

        fun dumpAllIntoPlayerInventory(player: ServerPlayer) {
            val size = containerSize;

            (0..size).forEach { slot ->
                val stack = getItem(slot)
                setItem(slot, ItemStack.EMPTY)
                player.inventory.placeItemBackInInventory(stack)
            }
        }

        fun insertItem(index: Int, stack: ItemStack, simulate: Boolean): ItemStack {
            this.addItem(stack).run { stack.shrink(this.count) }
            return stack
        }
    }

    override fun getItems(p0: Direction?): CommonStorage<ItemResource> {
        return WrappedVanillaContainer(handler);
    }
}
